components/builder-scheduler/src/data_store.rs

```diff
@@ -262,6 +261,43 @@ impl DataStore {
                            r#"CREATE OR REPLACE FUNCTION set_project_state_ident_v1 (pid bigint, jid bigint, state text, ident text) RETURNS void AS $$
                                   UPDATE projects SET project_state=state, job_id=jid, project_ident=ident, updated_at=now() WHERE id=pid;
                            $$ LANGUAGE SQL VOLATILE"#)?;
+
+        // Change the primary key for the groups table
+        migrator.migrate(
+            "scheduler",
+            r#"ALTER TABLE IF EXISTS groups DROP CONSTRAINT IF EXISTS groups_pkey;"#,
+        )?;
+        migrator.migrate(
+            "scheduler",
+            r#"CREATE SEQUENCE IF NOT EXISTS groups_id_seq;"#,
+        )?;
+        migrator.migrate("scheduler", r#"ALTER TABLE IF EXISTS groups ALTER COLUMN id SET DEFAULT next_id_v1('groups_id_seq');"#)?;
+        migrator.migrate(
+            "scheduler",
+            r#"ALTER TABLE IF EXISTS groups ADD PRIMARY KEY (id);"#,
+        )?;
+        migrator.migrate("scheduler",
+                     r#"CREATE OR REPLACE FUNCTION insert_group_v2 (
+                                project_names text[],
+                                project_idents text[]
+                                ) RETURNS SETOF groups AS $$
+                                    DECLARE
+                                        g groups % rowtype;
+                                    BEGIN
+                                        INSERT INTO groups (group_state)
+                                        VALUES ('Pending') RETURNING * INTO g;
+                                        RETURN NEXT g;
+
+                                        FOR i IN array_lower(project_names, 1)..array_upper(project_names, 1)
+                                        LOOP
+                                            INSERT INTO projects (owner_id, project_name, project_ident, project_state)
+                                            VALUES
+                                                (g.id, project_names[i], project_idents[i], 'NotStarted');
+                                        END LOOP;
+                                        RETURN;
+                                    END
+                                $$ LANGUAGE plpgsql VOLATILE
+                                "#)?;
```

Comment:
In general, it's better to implement function in SQL when possible, since the implementation is transparent to the planner. You can use the [unnest](https://www.postgresql.org/docs/current/static/functions-array.html) to convert your names array into a set, which you can treat just like a table, which means your `n` insert statements turn into a single statement.

You can also use [CTEs](https://www.postgresql.org/docs/9.6/static/queries-with.html) to put everything into a single SQL statement.

Hrmm... I presume that `project_names` and `project_idents` are the same length... is that confirmed when this function is called?